server/jetstream_api.go

```diff
@@ -687,36 +694,36 @@ func (js *jetStream) apiDispatch(sub *subscription, c *client, acc *Account, sub
 	}
 
 	// If we are here we have received this request over a non client connection.
-	// We need to make sure not to block. We will spin a Go routine per but also make
-	// sure we do not have too many outstanding.
-	if apiOut := atomic.AddInt64(&js.apiInflight, 1); apiOut > maxJSApiOut {
-		atomic.AddInt64(&js.apiInflight, -1)
-		ci, acc, _, msg, err := s.getRequestInfo(c, rmsg)
-		if err == nil {
-			resp := &ApiResponse{Type: JSApiOverloadedType, Error: NewJSInsufficientResourcesError()}
-			s.sendAPIErrResponse(ci, acc, subject, reply, string(msg), s.jsonResponse(&resp))
-		} else {
-			s.Warnf(badAPIRequestT, rmsg)
-		}
-		s.Warnf("JetStream API limit exceeded: %d calls outstanding", apiOut)
-		return
-	}
+	// We need to make sure not to block. We will send the request to a long-lived
+	// go routine.
+
+	// Copy the state. Note the JSAPI only uses the hdr index to piece apart the
+	// header from the msg body. No other references are needed.
+	s.jsAPIRoutedReqs.push(&jsAPIRoutedReq{jsub, sub, acc, subject, reply, copyBytes(rmsg), c.pa})
+}
+
+func (s *Server) processJSAPIRoutedRequests() {
+	defer s.grWG.Done()
 
-	// If we are here we can properly dispatch this API call.
-	// Copy the message and the client. Client for the pubArgs
-	// but note the JSAPI only uses the hdr index to piece apart
-	// the header from the msg body. No other references are needed.
-	// FIXME(dlc) - Should cleanup eventually and make sending
-	// and receiving internal messages more formal.
-	rmsg = copyBytes(rmsg)
+	s.mu.Lock()
+	queue := s.jsAPIRoutedReqs
 	client := &client{srv: s, kind: JETSTREAM}
-	client.pa = c.pa
+	s.mu.Unlock()
 
-	// Dispatch the API call to its own Go routine.
-	go func() {
-		jsub.icb(sub, client, acc, subject, reply, rmsg)
-		atomic.AddInt64(&js.apiInflight, -1)
-	}()
+	for {
+		select {
+		case <-queue.ch:
+			reqs := queue.pop()
+			for _, req := range reqs {
+				r := req.(*jsAPIRoutedReq)
+				client.pa = r.pa
```

Comment:
@derekcollison Want to bring to your attention that I am reusing a client that is created at the top of this function (we used to create one per request). From what I understand only the pubArg were really needed, so I think it is ok. If if there is a risk that the client's struct is used/modified by the icb() call, then let me know and we can create the client here for each request.

Reply:
Do we need to copy the pubArgs? Can they easily be overwritten?

I think its ok to re-use same client struct but we might need to deep copy pubArg.

Reply:
The code that this replace was doing a copy (not deep copy) of c.pa, however here, I agree that this is causing the copy to be done twice.

Regarding "we might need to deep copy pubArg": that was my concern when having the original code, see comment here: https://github.com/nats-io/nats-server/pull/1953#discussion_r586050261.

So now I am confused on what we should be doing here...

Reply:
Actually I think we already copy pubargs. So should be ok.

Reply:
Oh yes, I did not understand what you were asking here. The copy happens when pushing into the queue here: https://github.com/nats-io/nats-server/pull/3142/files#diff-9c075c47b0690a821ad35caa26308b69adeb132042f062eef75513db2a26638aR702

Reply:
I mean what it was referring to, not the []byte structs themselves.

Reply:
Since pubArgs is a struct, we already make a copy here:
```
s.jsAPIRoutedReqs.push(&jsAPIRoutedReq{jsub, sub, acc, subject, reply, copyBytes(rmsg), c.pa})
```
And, like I said, that bad part is that it is again making a copy (unfortunately) when doing:
```
client.pa = r.pa
```

Reply:
ok gotcha.